### PR TITLE
Add Proxy Tag Combination Script

### LINKED ISSUE #7300: Create script that can combine multiple asset tags for a user
The storage team works on features with the following high-level process:

image

Individual features are checked into a feature branch, which is then eventually checked into `main` just prior to shipping to `pypi`. Hotfixes to the previous release are normally serviced directly in `main`.

The problem is, when integrating any of the individual feature slices into the feature branch, **recordings** can conflict. This is due to the fact that the only reference to which recordings match up with the branch is the single `Tag` value in the `assets.json`. This doesn't give a good place to rationalize how to _combine_ the individual recordings from each `f` branch when merging into the `feature` branch.

Since we **definitely** don't want to be FORCED to rerecord each time we combine tags, here's what we're going to do.

We are going to start with a script that will utilize a `test-proxy` exe present on the PATH to:

1. accept a list of tags to be combined
2. cherry-pick the changes from each tag, attempt to combine
3. If there is a merge conflict, **leave the git directory alone and give it to the user so they can resolve the conflict themselves**
4. If there is _not_ a merge conflict, simply combine the recordings from each of the tags.
5. The order of the tags provided to the script will affect the _order which we combine the files as well_.

### PR DESCRIPTION
**Pull request by scbedd**
Resolves #7300 This script will be used to abstract the manual work that is combining multiple tags worth of recordings.

```powershell
merge-proxy-tags.ps1 "path/to/assets.json" "<tag1>" "<tag2>" "<tagN>"
```

This will checkout the current assets.json tag, then add N worth of `tags` to that changeset, via `cherry-pick`. On conflict, abandon _in place_ and allow resolution, followed by re-run of script to maintain the resolution.

TODO:

- [x] Works in place with an existing .assets folder in-repo
- [x] Combines multiple tags that aren't conflicting
- [x] Handle conflict scenario
  - Conflict scenario will utilize `azure/azure-sdk-assets-integration` repo and three manually created tags: `integration/example/storage_feature_addition1`, `integration/example/storage_feature_addition2`, `integration/example/storage_feature_addition2_conflict` which has been created and is ready to party on.
- [x] Touch a single file to allow the proxy to `push` without really knowing what changes are present.
- [x] Fresh documentation pass.
- [x] Confirm transition script still working

### CODE REVIEW

File: eng/common/testproxy/onboarding/common-asset-functions.ps1 [hunk 1]
```diff
@@ -0,0 +1,273 @@
+class Assets {
+  [string]$AssetsRepo = $DefaultAssetsRepo
+  [string]$AssetsRepoPrefixPath = ""
+  [string]$TagPrefix = ""
+  [string]$Tag = ""
+  Assets(
```
Comment (benbp) [inline eng/common/testproxy/onboarding/common-asset-functions.ps1:6]: Wow I did not know you could do this

File: eng/common/testproxy/onboarding/common-asset-functions.ps1 [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  # We're really only trying to get the language from the git remote
+  Write-Host "git remote -v"
+  [array] $remotes = & git remote -v
+  foreach ($line in $remotes) {
+    Write-Host "$line"
+  }
+
+  # Git remote -v returned "fatal: not a git repository (or any of the parent directories): .git"
+  # and the list of remotes will be null
+  if (-not $remotes) {
```
Comment (benbp) [inline eng/common/testproxy/onboarding/common-asset-functions.ps1:108]: Powershell doesn't store the error into the var?

File: eng/common/testproxy/onboarding/common-asset-functions.ps1 [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+  Assets(
+    [string]$AssetsRepoPrefixPath,
+    [string]$TagPrefix
+  ) {
+    $this.TagPrefix = $TagPrefix
+    $this.AssetsRepoPrefixPath = $AssetsRepoPrefixPath
+  }
+}
+
+class Version {
```
[the remark this answers is not in the document]
Comment (scbedd) [reply eng/common/testproxy/onboarding/common-asset-functions.ps1:15]: I figured if I was going to have a "common functions" class I may as well move everything that was in `generate-assets-json.ps1`.

I use `Get-Repo-Root` in the `merge-proxy-tags.ps1` script.

File: eng/common/testproxy/onboarding/common-asset-functions.ps1 [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+            return $false
+          }
+        }
+      }
+    }
+    return $true
+  }
+}
+
+Function Test-Exe-In-Path {
```
[the remark this answers is not in the document]
Comment (scbedd) [reply eng/common/testproxy/onboarding/common-asset-functions.ps1:60]: Yep, and it's `Get-Command`.

`where` is on `cmd` only.

File: eng/common/testproxy/onboarding/common-asset-functions.ps1 (hunk 2 above)
Comment (scbedd) [reply eng/common/testproxy/onboarding/common-asset-functions.ps1:108]: Nah that goes to stderr IIRC.

Comment (scbedd) [reply eng/common/testproxy/onboarding/common-asset-functions.ps1:108]: image

Confirmed.

File: eng/common/testproxy/onboarding/common-asset-functions.ps1 [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+    $CommandArgs = @(
+      "run --rm --name transition.test.proxy",
+      "-v `"${updatedDirectory}:/srv/testproxy`"",
+      "-e `"GIT_TOKEN=${token}`"",
+      "-e `"GIT_COMMIT_OWNER=${committer}`"",
+      "-e `"GIT_COMMIT_EMAIL=${email}`"",
+      $targetImage,
+      "test-proxy",
+      $CommandArgs
```
[the remark this answers is not in the document]
Comment (scbedd) [reply eng/common/testproxy/onboarding/common-asset-functions.ps1:221]: Yeah I'll fix this up. Confusing as hell.

File: eng/common/testproxy/scripts/tag-merge/merge-proxy-tags.ps1 [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+  4b. If the conflict was on the final tag, resolve the conflict (leaving it uncommitted tyvm), and test-proxy push, you're done.
+  4c. If the conflict was _not_ on the final tag, resolve the conflict, commit it, and then re-run this script with the SAME arguments as before.
+
+This script requires that test-proxy or azure.sdk.tools.testproxy should be on the PATH.
+
+.PARAMETER AssetsJson
+The script uses a target assets.json to resolve where specifically on disk the tag merging should take place.
+
+.PARAMETER TargetTags
+The set of tags whos contents should be merged into the targeted
```
Comment (semick-dev) [inline eng/common/testproxy/scripts/tag-merge/merge-proxy-tags.ps1:26]:
```suggestion
The set of tags whose contents should be merged combined. Any number of tags > 1 is allowed.
```

File: eng/common/testproxy/onboarding/common-asset-functions.ps1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (scbedd) [reply eng/common/testproxy/onboarding/common-asset-functions.ps1:221]: Actually array doesn't reference itself. I just reassign a string to another string value. (notice the -join) at the end of the line modifying CommandArgs.

I'm adding a comment for some clarity, but gonna leave it alone other than that.

**Review by benbp — APPROVED** [review]

**Pull request merged**